Security technote (#313)
apollographql/docs: src/content/technotes/TN0021-graph-security.mdx

id: TN0021
tags: [security, federation]
hidden: true
---

Generally speaking, many of the standard measures you would implement to reduce the attack surface of any API will apply to a GraphQL API in some way (the [OWASP Top Ten](https://owasp.org/www-project-top-ten/) list is a good place to start to understand some of the common risks). That said, there are some additional GraphQL-specific considerations you must make to limit your graph's exposure to potential threats. These threats are mostly related to denial-of-service (DoS) attacks and fall under the categories of API discoverability and malicious queries.

## API discoverability

One of the most important ways to protect a GraphQL API against would-be attackers who are trying to learn what vulnerabilities may exist within it is to limit API discoverability in production. While the inherent discoverability of a GraphQL API enhances developer experience when working locally, we typically don't want to offer these same affordances in a production environment for non-public APIs. The following sections explore some of the key ways to limit API discoverability.

### Turn off introspection in production

> Turning off introspection and access to GraphQL IDEs in production environments (as well as any publicly exposed staging environments) is the first step toward limiting the discoverability of the API because an introspection query is the quickest and easiest way for bad actors to learn about your schema. When using Apollo Router, introspection is turned off by default.

Reviewer: ```suggestion
Turning off introspection and access to integrated GraphQL IDEs in production environments (as well as any publicly exposed staging environments) is the first step toward limiting the discoverability of the API because an introspection query is the quickest and easiest way for bad actors to learn about your schema. When using Apollo Router, introspection is turned off by default.
```

Reviewer: Clarifying that things like Studio can work here could also be another avenue- but having GraphiQL or Sandbox on the production side is a no-go

### Obfuscate error details in production

Beyond introspection, many GraphQL servers attempt to enhance developer experience further by providing detailed error information in responses when something goes wrong. In a production environment, you’ll want to ensure that verbose error details are removed from API responses.

For example, a subgraph service such as Apollo Server provides the `exception.stacktrace` property under the `errors` key in a response, which can be useful while developing and debugging your server. In most cases, you shouldn't expose these stack trace details to public-facing clients, so Apollo Router will omit all error data by default.

You may wish to selectively expose some details of error messages sent back to clients in production environments, which may be done using a combination of [Apollo Router's `include_subgraph_errors` option](/router/configuration/subgraph-error-inclusion) and [Rhai scripts for response manipulation](https://www.apollographql.com/docs/router/customizations/rhai).

### Avoid autogenerating subgraph schemas

> Another important consideration when making a GraphQL API less discoverable is to avoid autogenerating schemas where possible, and particularly for the fields on the root operation types. There are many tools currently available that allow you to autogenerate a GraphQL schema quickly based on a handful of initial object type definitions in a schema or some existing database tables. While these approaches to schema generation may speed up initial API development, they also make it easier for bad actors to guess what kind of generic CRUD-related fields were autogenerated based on commonly used patterns.

Reviewer: I would either clarify why a subgraph specifically, or just nix the `subgraph` in the heading.

In the sections that follow, we will explore measures that can help mitigate the impact of malicious queries for any GraphQL API, such as limiting query depth and amount, paginating list fields where appropriate, validating and sanitizing data, setting timeouts, and guarding against batched query abuse. And for GraphQL APIs with third-party clients, we will also explore using query cost analysis to support rate-limiting.

### Limit query depth

GraphQL allows clients to traverse through a graph and express complex relationships between the nodes in an operation's selection set. But as far as backing data sources are concerned, this can quickly turn into too much of a good thing when there are no guardrails to restrict how deeply queries can be nested. For example:

```graphql
query DeepBlogQuery {
  author(id: 42) {
    posts {
      author {
        posts {
          author {
            posts {
              author {
                # and so on...
              }
            }
          }
        }
      }
    }
  }
}
```

> One of the easiest things to do is to guard against deeply nested queries such as this one is to set a maximum query depth. This can be done using an Apollo Router extension as demonstrated in [this Apollo Solutions repository](https://github.com/apollosolutions/router-basic-operation-cost/blob/main/src/plugins/basic_depth_limit.rs). And because an operation can specify multiple root fields, you may also consider limiting query breadth at the root level as well.

Reviewer: Do we want to recommend this out of the box? Feels like an endorsement to use right away - and not sure we'd want to promote as such unless blessed by the Router team.

Author: I can definitely remove it. We just don't have any other router-related action steps for this topic.

> ### Paginate fields where appropriate

Reviewer: Do we want to bring this up? Not sure how you would even begin to address this w/o custom plugins.

Author: Is this the same as your previous comment?

Reviewer: Effectively, yes, although I selected more lines than needed- only applied to 78 in this case. 

Probably worth making a decision on how we want to message this given some of the constraints.

Author: spoke offline; leaving as is until we have a more mature solution

### Batched requests

Batched requests are another potential attack vector for malicious queries. There are two different flavors of batching attacks to consider. The first threat is related to GraphQL's inherent ability to "batch" requests by allowing multiple root fields in an operation document:

```graphql
query {
  astronaut(id: "1") {
    name
  }
  second: astronaut(id: "2") {
    name
  }
  third: astronaut(id: "3") {
    name
  }
}
```

> Without any restrictions in place, clients could effectively enumerate through all nodes in a single request like the one above while slipping past other brute force protections. Limiting query breadth or using query cost analysis will help protect GraphQL APIs from this type of abuse.

Reviewer 2: ```suggestion
Without any restrictions in place, clients could effectively enumerate through all nodes in a single request like the one above while slipping past other brute force protections. Limiting query breadth or using query cost analysis can help protect a GraphQL API from this type of abuse.
```

Another form of batching occurs when a client sends batches of full operations in a single request, which can be helpful for performance reasons in some scenarios. **Apollo Router does not support this form of batching.** When operations are batched, Apollo Server receives an array of operations and sends back an array of responses to be parsed by the client (there's a batch link directly available in Apollo Client to facilitate this):

```graphql
[
  {
    “operationName”: "FirstAstronaut"
    “variables":{},
    "query":"query FirstAstronaut {\n  astronaut(id: \"1\") {\n    name\n  }\n}\n”
  },
  {
    “operationName”: "SecondAstronanut"
    “variables":{},
    "query":"query SecondAstronanut {\n  astronaut(id: \"2\") {\n    name\n  }\n}\n”
  },
  {
    “operationName”: "ThirdAstronaut"
    “variables":{},
    "query":"query ThirdAstronaut {\n  astronaut(id: \"3\") {\n    name\n  }\n}\n”
  }
]
```

> With batched operations, a key consideration to keep in mind is how they may impact rate limit calculations and query cost analysis to ensure that clients won't be able to cheat rate limits through race conditions.

Reviewer 2: ```suggestion
With batched operations, it's important to consider how an entire batch might impact rate limit calculations and query cost analysis to ensure that clients can't cheat rate limits through race conditions.
```

Finally, beyond batching of fields and operations, some forms of GraphQL-related batching can help mitigate DoS attacks and generally make your API more performant overall. Even with depth limiting in place, GraphQL queries can easily lead to exponential growth of requests to backing data sources. [Dataloaders](https://github.com/graphql/dataloader) are one way that you can help make as few requests as possible to backing data sources from resolver functions within the context of a single operation.xxw

## Security with managed federation

> Apart from protecting your GraphQL API from bad actors and locking down private data, you also need a window into how your API is being used and by whom to harden your GraphQL security posture. This is where a schema registry and observability tooling (such as [GraphOS](/graphos/)) come into play to help control who makes changes to your API and also monitor API usage and send alerts when something isn’t right.

Reviewer 2: ```suggestion
Apart from protecting your GraphQL API from bad actors and locking down private data, you also need a window into how your API is being used (and by whom) to harden your GraphQL security posture. This is where a schema registry and observability tooling (such as those provided by [GraphOS](/graphos/)) come into play to help control who makes changes to your API and also monitor API usage and send alerts when something isn’t right.
```

### Know who’s using your graph (and how)

> To enhance the utility of traces collected in your observability tooling, it's a best practice to require clients to identify themselves and that they consistently name operations. The web and mobile versions of Apollo Client make it easy to set custom headers for the name and version of a specific client to [segment traces and metrics in GraphOS by client](/graphos/metrics/client-awareness/). Other API clients can set the `apollographql-client-name` and `apollographql-client-version` request headers manually to provide client awareness as well. (As a bonus, client awareness also makes it easier to identify what clients may be impacted by a proposed breaking change to your API when running schema checks.)

Reviewer 2: ```suggestion
To enhance the utility of traces collected in your observability tooling, it's a best practice to require every client to identify itself _and_ assign a name to every operation it executes. The web and mobile versions of Apollo Client provide straightforward APIs for setting custom headers for a client's name and version. These help you [segment traces and metrics in GraphOS by client](/graphos/metrics/client-awareness/). Other API clients can set the `apollographql-client-name` and `apollographql-client-version` request headers manually to provide client awareness. (As a bonus, client awareness also helps you identify which clients might be impacted by a proposed breaking change to your API when running schema checks.)
```

> Additionally, tracing data in GraphOS can be used to monitor API performance and errors. You can also configure alerts to push notifications to your team when something goes wrong, whether it’s an increase in requests per minute, changes in your p50, p95, or p99 response times, or errors in operations run against your graph. For example, a notification about a sudden increase in the error percentage may indicate that a bad actor is trying to get around disabled introspection and learn about a graph's schema by rapidly guessing and testing different field names. And if you want to leverage error data outside of Studio as well, you can also use Apollo Router's support for [OpenTelemetry](/router/configuration/tracing) to integrate with other APM tools.

Reviewer 2: ```suggestion
Additionally, tracing data in GraphOS can help you monitor API performance and errors. You can configure alerts to push notifications to your team when something goes wrong, whether it’s an increase in requests per minute, changes in your p50, p95, or p99 response times, or errors in operations run against your graph. For example, a notification about a sudden increase in the error percentage might indicate that a bad actor is trying to circumvent disabled introspection and learn about a graph's schema by rapidly guessing and testing different field names. And if you want to leverage error data outside of GraphOS as well, you can also use the Apollo Router's support for [OpenTelemetry](/router/configuration/tracing) to integrate with other APM tools.
```

### Restrict write access to your graph

> Just as important as it is to set some limitations around how the outside world interacts with your graph, internal access to your GraphQL API should be managed thoughtfully as well. GraphOS provides both [graph API keys and personal API keys](/graphos/api-keys/) to restrict access to the graphs within an organization. It also supports SSO integration and different [member roles](/graphos/org/members/#organization-wide-member-roles) so that team members can be assigned appropriate permissions when contributing the graph.

Reviewer 2: ```suggestion
You should manage _internal_ access to your supergraph as thoughtfully as you manage communication from external clients. GraphOS provides both [graph API keys and personal API keys](/graphos/api-keys/) to restrict access to the graphs within an organization. It also supports SSO integration and different [member roles](/graphos/org/members/#organization-wide-member-roles) so that team members can be assigned appropriate permissions when contributing the graph.
```
